semtypes/src/test/java/io/ballerina/types/CellTypeTest.java

```diff
@@ -0,0 +1,364 @@
+/*
+ *  Copyright (c) 2024, WSO2 LLC. (https://www.wso2.com).
+ *
+ *  WSO2 LLC. licenses this file to you under the Apache License,
+ *  Version 2.0 (the "License"); you may not use this file except
+ *  in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing,
+ *  software distributed under the License is distributed on an
+ *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ *  KIND, either express or implied. See the License for the
+ *  specific language governing permissions and limitations
+ *  under the License.
+ */
+package io.ballerina.types;
+
+import io.ballerina.types.subtypedata.CellSubtype;
+import org.testng.Assert;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+import static io.ballerina.types.CellAtomicType.CellMutability.CELL_MUT_LIMITED;
+import static io.ballerina.types.CellAtomicType.CellMutability.CELL_MUT_NONE;
+import static io.ballerina.types.CellAtomicType.CellMutability.CELL_MUT_UNLIMITED;
+
+/**
+ * Tests subtyping rules of cell basic type.
+ *
+ * @since 2201.10.0
+ */
+public class CellTypeTest {
+
+    Context ctx;
+
+    public enum Relation {
+        EQUAL("="),
+        SUBTYPE("<"),
+        NO_RELATION("<>");
+
+        final String value;
+
+        Relation(String value) {
+            this.value = value;
+        }
+    }
+
+    @BeforeClass
+    public void beforeClass() {
+        ctx = Context.from(new Env());
+    }
+
+    private CellSemType cell(SemType ty, CellAtomicType.CellMutability mut) {
+        return CellSubtype.cellContaining(ctx.env, ty, mut);
+    }
+
+    private SemType tuple(SemType ty) {
+        return SemTypes.tuple(ctx.env, new SemType[]{ty});
+    }
+
+    private void assertSemTypeRelation(SemType t1, SemType t2, Relation relation) {
+        Relation actual = getSemTypeRelation(t1, t2);
+        Assert.assertEquals(actual, relation);
+    }
+
+    private Relation getSemTypeRelation(SemType t1, SemType t2) {
+        boolean s1 = Core.isSubtype(ctx, t1, t2);
+        boolean s2 = Core.isSubtype(ctx, t2, t1);
+        if (s1 && s2) {
+            return Relation.EQUAL;
+        } else if (s1) {
+            return Relation.SUBTYPE;
+        } else if (s2) {
+            throw new IllegalStateException("'>' relation found which can be converted to a '<' relation");
+        } else {
```

Comment:
Wedon't need this else statement in here, since we returning in other conditions

Reply:
For subtyping between two types say, T1 and T2. There are only 4 scenarios. 
1. T1 = T2 
2. T1 <: T2
3. T2 <:  T1 (Can be written in form (2). Hence, the error.)
4. T1 <> T2

Here with this style branching for 4 cases is clear and easy to understand. WDYT?

Reply:
+1